lib/vtls/mbedtls.c

```diff
@@ -658,20 +616,45 @@ static CURLcode mbed_connect_step1(struct Curl_cfilter *cf,
   }
 
   if(ssl_cert_blob) {
-    /* Unfortunately, mbedtls_x509_crt_parse() requires the data to be null
-       terminated even when provided the exact length, forcing us to waste
-       extra memory here. */
-    unsigned char *newblob = Curl_memdup0(ssl_cert_blob->data,
-                                          ssl_cert_blob->len);
-    if(!newblob)
-      return CURLE_OUT_OF_MEMORY;
-    ret = mbedtls_x509_crt_parse(&backend->clicert, newblob,
-                                 ssl_cert_blob->len + 1);
-    curlx_free(newblob);
+#ifdef MBEDTLS_PEM_PARSE_C
+    /* if DER or a null-terminated PEM just process using
+    * mbedtls_x509_crt_parse().
+    * */
+    if((ssl_cert_type && strcmp(ssl_cert_type, "DER") == 0)
+       || ((char *)(ssl_cert_blob->data))[ssl_cert_blob->len - 1] == '\0'
+      ) {
+      ret = mbedtls_x509_crt_parse(&backend->clicert,
+                                   ssl_cert_blob->data,
+                                   ssl_cert_blob->len);
+    }
+    else /* they say it is PEM and it is not null-terminated */
+    {
+      /* Unfortunately, mbedtls_x509_crt_parse() requires the data to
+         be null-terminated if the data is PEM encoded (even when
+         provided the exact length). The function accepts PEM or DER
+         formats, but we cannot assume if the user passed in a PEM
+         format cert that it is null-terminated. */
+      unsigned char *newblob = Curl_memdup0(ssl_cert_blob->data,
+                                            ssl_cert_blob->len);
+      if(!newblob)
+        return CURLE_OUT_OF_MEMORY;
+      ret = mbedtls_x509_crt_parse(&backend->clicert, newblob,
+                                   ssl_cert_blob->len + 1);
+      curlx_free(newblob);
+    }
+#else
+    /* DER encoded certs do not need to be null terminated
+    because it is a binary format. So if we are not compiling
+    with PEM_PARSE we can avoid the extra memory copies
+    altogether. */
+    ret = mbedtls_x509_crt_parse_der(&backend->clicert,
+                                     ssl_cert_blob->data,
+                                     ssl_cert_blob->len);
+#endif
 
     if(ret) {
       mbedtls_strerror(ret, errorbuf, sizeof(errorbuf));
-      failf(data, "mbedTLS: error reading client cert data %s: (-0x%04X) %s",
+      failf(data, "mbedTLS: error reading SSL cert blob %s: (-0x%04X) %s",
```

Comment:
I personally think *client cert* is more precisely explaining which cert it is that causes the problem...

Reply:
Okie doke. Fine with me.